Add support for question README

mylesw27 opened the pull request:
Closes #5357. This PR will allow us to support and display a README.md file for questions. This will check if a README is present for the question and if so, display it in a panel above the question in the question preview page. 
![Screenshot 2025-06-02 at 10 03 09 AM](https://github.com/user-attachments/assets/32c12447-1650-44ff-8479-c6fbceae0eac)

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionPreview/instructorQuestionPreview.html.ts:
```diff
@@ -84,6 +86,27 @@ export function InstructorQuestionPreview({
+                    <h2>README</h2>
+                  </div>
+                  <div class="card-body collapse show" id="readme-card-body">
+                    <div class="markdown-body">${unsafeHtml(readmeHtml)}</div>
```
IMO you'd benefit from styles similar to this here:

https://github.com/PrairieLearn/PrairieLearn/blob/09b9386f2e32621310fb2f730435d555dec39870/apps/prairielearn/src/pages/studentAssessment/studentAssessment.html.ts#L31-L33

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionPreview/instructorQuestionPreview.ts:
```diff
@@ -112,6 +115,15 @@ router.get(
     } else if (aiGradingPreviewEnabled) {
       renderSubmissionSearchParams.set('ai_grading_preview', 'true');
     }
+    const questionReadmePath = path.join(
+      path.join(res.locals.course.path, 'questions', res.locals.question.qid, 'README.md'),
+    );
+    const questionReadmeExists = await fs.pathExists(questionReadmePath);
+    let readmeHtml = '';
+    if (questionReadmeExists) {
+      const readme = await fs.readFile(questionReadmePath, 'utf8');
+      readmeHtml = await markdownToHtml(readme, { allowHtml: false });
+    }
```
If there are indeed file system errors, we'd want that to bubble up so that we as admins would be aware of that. The current lack of explicit error handling is best for us.

nwalters512 commented on testCourse/questions/addNumbers/README.md:
```diff
@@ -0,0 +1,3 @@
+## Add numbers
+
```
I'd drop the heading here:

```suggestion
```

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionPreview/instructorQuestionPreview.ts:
```diff
@@ -112,6 +115,15 @@ router.get(
     } else if (aiGradingPreviewEnabled) {
       renderSubmissionSearchParams.set('ai_grading_preview', 'true');
     }
+    const questionReadmePath = path.join(
+      path.join(res.locals.course.path, 'questions', res.locals.question.qid, 'README.md'),
+    );
+    const questionReadmeExists = await fs.pathExists(questionReadmePath);
+    let readmeHtml = '';
+    if (questionReadmeExists) {
+      const readme = await fs.readFile(questionReadmePath, 'utf8');
+      readmeHtml = await markdownToHtml(readme, { allowHtml: false });
+    }
```
This is going to cause problems in production, where the question preview page is served from chunk servers where `res.locals.course.path` doesn't exist on disk.

You'll want to use `getRuntimeDirectoryForCourse(...)` to get a course directory. Locally, it'll point at just the course repository itself, but on chunk servers, it'll point to the directory where we've constructed the course data from chunks.

You can see [`apps/prairielearn/src/pages/clientFilesQuestion/clientFilesQuestion.ts`](https://github.com/PrairieLearn/PrairieLearn/blob/master/apps/prairielearn/src/pages/clientFilesQuestion/clientFilesQuestion.ts) as an example of something that's doing the right thing to read files out of a question's directory.

You shouldn't need an explicit `ensureChunks(...)` call here, as the question rendering process above (`getAndRenderVariant(...)`) will already have done that. You can leave a comment to that effect.

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionPreview/instructorQuestionPreview.html.ts:
```diff
@@ -84,6 +86,27 @@ export function InstructorQuestionPreview({
+                  <div class="card-header bg-secondary text-white d-flex align-items-center">
+                    <button
+                      class="btn btn-secondary btn-xs show-hide-btn"
+                      data-bs-toggle="collapse"
```
The collapse/expand animation here is a little jerky, it jumps around at the start/end of the animation. I'm not 100% sure what's causing that, but you can maybe look at the submission panels as examples of something that's working smoothly.

mylesw27 replied:
This was a strange one. Needed to add an extra `<div>` wrapping the card body to handle the `.collapsing` class that Bootstrap applies. Now works smoothly after [316b1e7](https://github.com/PrairieLearn/PrairieLearn/pull/12093/commits/316b1e7de64df46142185a085f4d69a442e1990b).

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionPreview/instructorQuestionPreview.html.ts:
```diff
@@ -84,6 +86,27 @@ export function InstructorQuestionPreview({
+                      aria-expanded="true"
+                      aria-controls="#readme-card-body"
+                    >
+                      <i class="fa fa-angle-up fa-fw expand-icon"></i>
```
I'd be inclined to put this button on the right side of the card header to mirror the submission panels.

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionPreview/instructorQuestionPreview.html.ts:
```diff
@@ -84,6 +86,27 @@ export function InstructorQuestionPreview({
+          ${readmeHtml
+            ? html`
+                <div class="card mb-3">
+                  <div class="card-header bg-secondary text-white d-flex align-items-center">
```
I might be inclined to use a `light` background here instead of `secondary`. There are already a lot of things on this page with quite a bit of visual weight, I think we can de-emphasize this.

nwalters512 commented on apps/prairielearn/src/pages/instructorQuestionPreview/instructorQuestionPreview.html.ts:
```diff
@@ -84,6 +86,27 @@ export function InstructorQuestionPreview({
+                    <h2>README</h2>
+                  </div>
+                  <div class="card-body collapse show" id="readme-card-body">
+                    <div class="markdown-body">${unsafeHtml(readmeHtml)}</div>
```
Is the `markdown-body` class actually used by anything?

reteps replied:
Can that be done with bootstrap classes instead of a custom class?

nwalters512 commented on testCourse/questions/addNumbers/README.md:
```diff
@@ -0,0 +1 @@
+This is a basic question to test the addition of two integers.
```
A heading isn't useful in question README files. The question title is already shown on the page.

trombonekenny commented:
> > We should mark that the README is just visible to instructors, maybe similar to what we do with the Staff Information card on that page? I'm worried instructors might see and start using it for student instructions, which aren't shown to them.
> > Many instructors seem to base the question preview page as what the student sees too - which isn't always true.
> 
> @trombonekenny I added a footer in [e56938b](https://github.com/PrairieLearn/PrairieLearn/pull/12093/commits/e56938b66a02ce11d5e5775fd89d1da567478e5e). Does something like this seem like what we want here?

It's good - if we didn't want a footer, we could put it in the header instead. Something like `README.md (only visible to instructors)`

jonatanschroeder commented on apps/prairielearn/src/pages/instructorQuestionPreview/instructorQuestionPreview.ts:
```diff
@@ -117,11 +117,16 @@ router.get(
     const questionReadmePath = path.join(
       path.join(coursePath, 'questions', res.locals.question.qid, 'README.md'),
     );
+    // We do not not need an explicit `ensureChunks()` call here as the `getAndRenderVariant()` call
```
```suggestion
    // We do not need an explicit `ensureChunks()` call here as the `getAndRenderVariant()` call
```

mylesw27 commented on apps/prairielearn/src/pages/instructorQuestionPreview/instructorQuestionPreview.ts:
```diff
@@ -109,6 +114,22 @@ router.get(
+    const coursePath = getRuntimeDirectoryForCourse(question_course);
+    const questionReadmePath = path.join(
+      path.join(coursePath, 'questions', res.locals.question.qid, 'README.md'),
+    );
```
Added a line here in [cc91914](https://github.com/PrairieLearn/PrairieLearn/pull/12093/commits/cc91914bd71bf3b2a5599674aefcdaca675b4b95).

mylesw27 commented on apps/prairielearn/src/pages/instructorQuestionPreview/instructorQuestionPreview.html.ts:
```diff
@@ -84,6 +91,33 @@ export function InstructorQuestionPreview({
+            ? html`
+                <div class="card mb-3">
+                  <div
+                    class="card-header bg-light text-dark d-flex align-items-center submission-header"
```
Nice catch. Removed those unneeded classes in [cc91914](https://github.com/PrairieLearn/PrairieLearn/pull/12093/commits/cc91914bd71bf3b2a5599674aefcdaca675b4b95).

mylesw27 commented on apps/prairielearn/src/pages/instructorQuestionPreview/instructorQuestionPreview.html.ts:
```diff
@@ -84,6 +91,33 @@ export function InstructorQuestionPreview({
+                      aria-expanded="true"
+                      aria-controls="#readme-card-body"
+                    >
+                      <i class="fa fa-angle-up fa-fw ms-1 expand-icon"></i>
```
It is possible and I have updated `local.css` where these values are being set to now be "Show" and "Hide". This also impacts the submission panel but I believe we use `Sentence casing` for all of our buttons so I think this update should be fine.

miguelaenlle commented on apps/prairielearn/assets/scripts/question.ts:
```diff
@@ -32,6 +32,19 @@ onDocumentReady(() => {
     loadPendingSubmissionPanel(e.currentTarget, false);
   });
 
+  $<HTMLDivElement>('.js-submission-body').on('show.bs.collapse', (e) => {
+    e.target
+      .closest('.card')
+      ?.querySelector<HTMLDivElement>('.submission-header')
+      ?.classList.remove('border-bottom-0');
+  });
+  $<HTMLDivElement>('.js-submission-body').on('hidden.bs.collapse', (e) => {
+    e.target
+      .closest('.card')
+      ?.querySelector<HTMLDivElement>('.submission-header')
+      ?.classList.add('border-bottom-0');
+  });
```
Nit: I agree with using vanilla JS for consistency in this file and also refactoring the `$<HTMLDivElement>('.js-submission-body.render-pending')` jQuery code above, though I won't block on this change.

mylesw27 replied:
I agree. Updated to use vanilla JS here in [614eb07](https://github.com/PrairieLearn/PrairieLearn/pull/12093/commits/614eb07c2163641284175e47920547d90c9d7509).

miguelaenlle reviewed (approved):
I tested the question README locally on a variety of questions, and it works very well! My only comment was a minor nit related to the CodeRabbit suggestion.

mylesw27 commented on apps/prairielearn/assets/scripts/question.ts:
```diff
@@ -32,6 +32,23 @@ onDocumentReady(() => {
   });
 
+  document.addEventListener('show.bs.collapse', (e) => {
+    if ((e.target as HTMLElement)?.classList.contains('js-submission-body')) {
```
That's a good point. I was just trying to reuse existing behavior. Renamed these classes to how they are actually being used in [0b98d77](https://github.com/PrairieLearn/PrairieLearn/pull/12093/commits/0b98d77ca8b58aadb60cb66f944d7d7202d0295e). I am now using the suggested `js-collapsible-card-body` and `collapsible-card-header` for these classes. I replaced most of the usage with the exception of `js-submission-body` which as you pointed out, is used for different things. I added `js-collapsible-card-body` here and left the existing `js-submission-body`.

nwalters512 commented:
This is mostly ready to go! As discussed in meetings today, we're concerned about the panel showing at full height on page load. We want to try implementing something like the "expand" experience that we added in https://github.com/PrairieLearn/PrairieLearn/pull/11223.

mylesw27 commented:
> This is mostly ready to go! As discussed in meetings today, we're concerned about the panel showing at full height on page load. We want to try implementing something like the "expand" experience that we added in #11223.

I added the "expand" button and limited the readme to 150 pixels on page load. The display now renders like this on page load:

![Screenshot 2025-07-09 at 10 02 23 AM](https://github.com/user-attachments/assets/d1bbbd8b-b206-46c3-9758-69d0dd2a2cc8)

nwalters512 reviewed (commented):
This is more of a nit than anything. I'll let @reteps offer final approval!

reteps commented:
In [2031edc](https://github.com/PrairieLearn/PrairieLearn/pull/12093/commits/2031edc81d63953dd65642ccc1ef6c668e46a579) I add the ability to collapse again, and put the gradient below the 'collapse' button.
